I've added an intentional exception in the code:
```python
         log.debug(f"Making request: {method} {url} with data: {data}")
         try:
             result = request(
                 method=method,
                 url=url,
                 json=data,
                 headers=headers,
             )
             raise Exception("This is intentional")  # Simulated exception for testing retries
         except Exception as e:
             log.warning("Simulated exception: %s", str(e))
             raise

```
 to test how it behaves and as expected:
```bash
mburcul@node-01.he-eu-hel1.ci.gitea:~ % ./mirror.py -l info codex-storage --log-level DEBUG
2025-01-14 11:01:22 - INFO - Gitea URL: http://127.0.0.1:3000/api/v1/
2025-01-14 11:01:22 - DEBUG - Making request: GET http://127.0.0.1:3000/api/v1/version with data: None
2025-01-14 11:01:22 - WARNING - Simulated exception: This is intentional
2025-01-14 11:01:23 - DEBUG - Making request: GET http://127.0.0.1:3000/api/v1/version with data: None
2025-01-14 11:01:23 - WARNING - Simulated exception: This is intentional
2025-01-14 11:01:25 - DEBUG - Making request: GET http://127.0.0.1:3000/api/v1/version with data: None
2025-01-14 11:01:26 - WARNING - Simulated exception: This is intentional
2025-01-14 11:01:30 - DEBUG - Making request: GET http://127.0.0.1:3000/api/v1/version with data: None
2025-01-14 11:01:30 - WARNING - Simulated exception: This is intentional
2025-01-14 11:01:38 - DEBUG - Making request: GET http://127.0.0.1:3000/api/v1/version with data: None
2025-01-14 11:01:38 - WARNING - Simulated exception: This is intentional
Traceback (most recent call last):
  File "/home/mburcul/./mirror.py", line 24, in wrapper
    return func(*args, **kwargs)
  File "/home/mburcul/./mirror.py", line 53, in _request
    raise Exception("This is intentional")  # Simulated exception for testing retries
Exception: This is intentional

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/mburcul/./mirror.py", line 224, in <module>
    main()
  File "/home/mburcul/./mirror.py", line 192, in main
    log.info('Gitea Version: %s', ga.version())
  File "/home/mburcul/./mirror.py", line 65, in version
    return self._request('GET', 'version').json()['version']
  File "/home/mburcul/./mirror.py", line 28, in wrapper
    raise Exception(f"Max retries of function {func.__name__} exceeded") from e
Exception: Max retries of function _request exceeded
```

--- a/files/mirror.py
+++ b/files/mirror.py
@@ -1,5 +1,6 @@
 #!/usr/bin/env python3
 
+import time
 import argparse
 import logging as log
 from re import match
@@ -13,6 +14,23 @@ from requests import request, exceptions
 log.getLogger('urllib3').setLevel(log.WARN)
 log.getLogger('github').setLevel(log.WARN)
 
+def retry(max_retries, wait_time):
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            retries = 0
+            delay = wait_time
+            while retries < max_retries:
+                try:
+                    return func(*args, **kwargs)
+                except Exception as e:
+                    retries += 1
+                    if retries >= max_retries:
+                        raise Exception(f"Max retries of function {func.__name__} exceeded") from e
+                    time.sleep(delay)
+                    delay *= 2
+        return wrapper
+    return decorator
+
 class Gitea:
 
     def __init__(self, url, token, gh_token):
@@ -20,22 +38,17 @@ class Gitea:
         self.token = token
         self.gh_token = gh_token
 
+    @retry(max_retries=5, wait_time=1)
     def _request(self, method, path, data=None):
         url = urljoin(self.url, path)
         headers = { 'Authorization': 'token ' + self.token }
-        try:
-            result = request(
-                method=method,
-                url=url,
-                json=data,
-                headers=headers,
-            )
-        except exceptions.ConnectionError as e:
-            log.error('ConnectionError on url %s: %s', url, e)
-            return None
-        if result.status_code == 404:
-            result.raise_for_status()
-        if result.status_code not in [200, 201, 404]:
+        result = request(
+            method=method,
+            url=url,
+            json=data,
+            headers=headers,
+        )
+        if result.status_code not in [200, 201]:
             log.error('Request failed: %s %s (%d)', method, url, result.status_code)
             log.error('Error: %s', result.text)
             result.raise_for_status()

